src/storage/src/mem_table.rs

```diff
@@ -182,3 +221,395 @@ impl KeyOp {
         }
     }
 }
+
+#[try_stream(ok = StateStoreIterItem, error = StorageError)]
+async fn merge_stream<'a>(
+    mem_table_iter: impl Iterator<Item = (&'a Bytes, &'a KeyOp)> + 'a,
+    inner_stream: impl StateStoreReadIterStream,
+    table_id: TableId,
+    epoch: u64,
+) {
+    let inner_stream = inner_stream.peekable();
+    pin_mut!(inner_stream);
+
+    let mut mem_table_iter = mem_table_iter.fuse().peekable();
+
+    loop {
+        match (inner_stream.as_mut().peek().await, mem_table_iter.peek()) {
+            (None, None) => break,
+            // The mem table side has come to an end, return data from the shared storage.
+            (Some(_), None) => {
+                let (key, value) = inner_stream.next().await.unwrap()?;
+                yield (key, value)
+            }
+            // The stream side has come to an end, return data from the mem table.
+            (None, Some(_)) => {
+                let (key, key_op) = mem_table_iter.next().unwrap();
+                match key_op {
+                    KeyOp::Insert(value) | KeyOp::Update((_, value)) => {
```

Comment:
If there's no such key in the storage, should we throw an `Inconsistent` error for `Update` here?

Reply:
I think this will be done in `StateStore::flush`